### PR TITLE
Fix lr scheduler state not being dumped to checkpoint in deepspeed strategy

### LINKED ISSUE #11188: lr scheduler state not saved with DeepSpeed plugin
## 🐛 Bug

When using deepspeed plugin, the learning rate scheduler state is not saved. The key `lr_scheduler` in `mp_rank_00_model_states.pt` model file is None. The bug causes the resuming from checkpoint to behave unexpected, where it should continue from previous lr scheduler state but instead lr scheduler start from scratch. This issue happens for most pytorch_lightning release versions, but the issue does not happen for some commits, e.g. commit 86c033572da084fd85d4d4c2d637db18d2d431ec.

### To Reproduce

```
import os
import torch
from torch.utils.data import Dataset, DataLoader
import pytorch_lightning as pl
from pytorch_lightning import LightningModule, Trainer
from pytorch_lightning.callbacks import ModelCheckpoint
from pytorch_lightning.plugins import DeepSpeedPlugin
from deepspeed.ops.adam import FusedAdam
from transformers.optimization import AdamW, get_scheduler

class RandomDataset(Dataset):

    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):

    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def train_dataloader(self):
        return DataLoader(RandomDataset(32, 64), batch_size=2)
    
    def test_dataloader(self):
        return DataLoader(RandomDataset(32, 64), batch_size=2)

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("train_loss", loss)
        return loss

    def validation_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("valid_loss", loss)
        return loss

    def test_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("test_loss", loss)
        return loss

    def configure_optimizers(self):
        no_decay = ["bias"]
        params_decay = [
            p for n, p in self.named_parameters() if not any(nd in n for nd in no_decay)
        ]
        params_nodecay = [
            p for n, p in self.named_parameters() if any(nd in n for nd in no_decay)
        ]
        optim_groups = [
            {
                "params": params_decay,
                "weight_decay": 0.01,
            },
            {"params": params_nodecay, "weight_decay": 0.0},
        ]
        optimizer = torch.optim.SGD(optim_groups, lr=0.1)
        scheduler = get_scheduler("linear", optimizer, num_warmup_steps=20, num_training_steps=200)
        return ([optimizer], [{"scheduler": scheduler, "interval": "step", "frequency": 1}])

def run():
    train_data = DataLoader(RandomDataset(32, 64), batch_size=4)
    val_data = DataLoader(RandomDataset(32, 64), batch_size=4)
    test_data = DataLoader(RandomDataset(32, 64), batch_size=4)

    model = BoringModel()
    checkpoint_callback = ModelCheckpoint(
        dirpath='tests4/checkpoints',
        save_last=True,
        every_n_train_steps=10,
    )
    trainer = Trainer(
        default_root_dir=os.getcwd(),
        gpus=-1,
        limit_train_batches=1.0,
        limit_val_batches=1,
        num_sanity_val_steps=0,
        precision=16,
        accelerator='deepspeed',
        max_steps=100,
        log_every_n_steps=1,
        plugins=[DeepSpeedPlugin(stage=2)],
        weights_summary=None,
        callbacks=[checkpoint_callback, pl.callbacks.LearningRateMonitor(logging_interval="step")],
        #resume_from_checkpoint='tests4/checkpoints/epoch=3-step=99.ckpt',
    )
    trainer.fit(model)
    trainer.test(model)

    # test lr_scheduler state
    model_state_path = "tests4/checkpoints/epoch=3-step=99.ckpt/global_step100/mp_rank_00_model_states.pt"
    model_state = torch.load(model_state_path, map_location="cpu")
    print(model_state["lr_schedulers"])


if __name__ == '__main__':
    run()
```

Just run `python run.py`. For latest release version 1.5.6 and many other versions including current master branch, it produces None, meaning lr_scheduler state is not saved. For commit id 86c033572da084fd85d4d4c2d637db18d2d431ec, it produces the correct state:
```
{'base_lrs': [0.1, 0.1], 'last_epoch': 47, '_step_count': 48, 'verbose': False, '_get_lr_called_within_step': False, '_last_lr': [0.05888888888888889, 0.05888888888888889], 'lr_lambdas': [None, None]}
```

Do first run with the code exactly as above. Do second run by changing max_steps from 100 to 200, and uncomment `#resume_from_checkpoint` line. The two runs gives the tensorboard curves of resuming training. For pytorch lightning version 1.5.6 and many others, it produces wrong lr curve:
Screen Shot 2021-12-20 at 5 19 20 PM

However, the correct resuming curve (like the one produced with commit id 86c033572da084fd85d4d4c2d637db18d2d431ec) should be
Screen Shot 2021-12-20 at 5 16 57 PM

### Expected behavior

The lr_scheduler state should be saved properly in order to resume from checkpoint correctly. Could you clarify if we should handle lr_scheduler state loading or saving by ourselves, or it should be properly handled by pytorch_lightning? If it should be handled by pytorch_lightning, do you know how to resolve the bug?

cc @tchaton @SeanNaren @awaelchli

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Part of #11188

The added assertion fails on master.

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

Part of #1 (it's a lie, this is just here to avoid noisy GitHub bot)

cc @borda @SeanNaren @awaelchli @rohitgr7

### CODE REVIEW

**Review by SeanNaren — APPROVED** [review]
Thanks for fixing this @awaelchli. I assume based on the logic, the scheduler will be loaded if found in the state dict elsewhere in the code

File: pytorch_lightning/strategies/deepspeed.py [hunk 1]
```diff
@@ -741,7 +741,7 @@ def save_checkpoint(self, checkpoint: Dict, filepath: _PATH) -> None:
             )
         # Use deepspeed's internal checkpointing function to handle partitioned weights across processes
         # dump states as a checkpoint dictionary object
-        _exclude_keys = ["state_dict", "optimizer_states", "lr_schedulers"]
+        _exclude_keys = ["state_dict", "optimizer_states"]
```
Comment (carmocca) [inline pytorch_lightning/strategies/deepspeed.py:744]: If the `lr_schedulers` weren't getting saved, why didn't this test fail? It checks that the `lr_schedulers` key exists in the saved checkpoint

https://github.com/PyTorchLightning/pytorch-lightning/blob/master/tests/models/test_hooks.py#L496

**Review by carmocca — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

**Review by rohitgr7 — APPROVED** [review]

File: pytorch_lightning/strategies/deepspeed.py (hunk 1 above)
Comment (awaelchli) [reply pytorch_lightning/strategies/deepspeed.py:744]: Ah, I didn't know this hook test uses deepspeed. Thanks! 

It's because the entry is a list and it is empty if no scheduler is defined/saved. And so this matches ANY.

Comment (awaelchli) [reply pytorch_lightning/strategies/deepspeed.py:744]: and `on_save_checkpoint` where the test asserts the contents gets called before the `save_checkpoint` call (where deepspeed dropped the key)

**Comment by tridao** [comment]
> Thanks for fixing this @awaelchli. I assume based on the logic, the scheduler will be loaded if found in the state dict elsewhere in the code

I'm not sure that's the case, since the Deepspeed plugin has `lightning_restore_optimizer_and_schedulers` set to False.
Would the [checkpoint_connector](https://github.com/PyTorchLightning/pytorch-lightning/blob/7eab379da2fdca542849ed4ad313d0851c2271e3/pytorch_lightning/trainer/connectors/checkpoint_connector.py#L261) ignore the lr_schedulers key when restoring?
```
def restore_optimizers_and_schedulers(self) -> None:
        """Restores the optimizers and learning rate scheduler states from the pre-loaded checkpoint."""
        if not self._loaded_checkpoint or not self.trainer.strategy.lightning_restore_optimizer_and_schedulers:
            return
```

**Comment by awaelchli** [comment]
Yes you are right, this needs to be taken care of too.

**Comment by eelxpeng** [comment]
Right. I tried with current commits. Though `lr_schedulers` are saved, the `lr_scheduler` state will not be restored.
